### PR TITLE
[browser][MT] deputy thread

### PR DESCRIPTION
**Pull request by pavelsavara**
## Threading modes
- described in `multi-threading.md`
- new `JSProxyContext.ThreadBlockingMode` flowing from `config.`
    - influences `ThrowOnBlockingWaitOnJSInteropThread` behavior
- new `JSProxyContext.ThreadInteropMode`
    - influences sychronous calls to `JSImport` and `JSExport`
    - new `JSProxyContext.IsPendingSynchronousCall`
    - new `BeforeSyncJSExport` and `AfterSyncJSExport` to recognize start/end of incoming sync `JSImport` call.
    - new `mono_wasm_invoke_jsexport_sync_send_cb` calls it
    - `mono_wasm_install_js_worker_interop_wrapper` helps with the setup of it.
    - new places throw PNSE when the rules are breached
    - `args.CallerNativeTID` to see if the `JSImport` caller is the same thread. 
        - only set `runtimeHelpers.isPendingSynchronousCall` when are same
    - `get_caller_native_tid` to read that
- new `JSProxyContext.MainThreadingMode`
    - could be used to switch off deputy thread. Not recommended.
- new enums `JSThreadBlockingMode` and `JSThreadInteropMode` and `MainThreadingMode`

## Deputy creation
- new `mono_wasm_create_deputy_thread`
    - create deputy thread and start mono VM on it
- new `mono_wasm_register_ui_thread`
    - and later attach the UI thread to Mono. We hope to avoid this in the future.

## testing
- there is new flavor of `System.Runtime.InteropServices.JavaScript.BackgroundExec.Tests.csproj` 
    - disabled for ST
    - to run tests which really need to run from non-JS thread.
    - new env variable `IsWasmBackgroundExec` passed to unit tests

### CODE REVIEW

**Comment by pavelsavara** [comment]
Testing with Blazor

This is call stack for `renderBatch`, it shows no managed code
image

This is call stack for HTTP client called from thread pool, it shows no managed code
image

This is `endInvokeJSFromDotNet` on deeper stack and `invokeJSJson` inside of that. This needs to be improved.
image


This is `endInvokeJSFromDotNet` on deeper stack and `renderBatch` inside of that. This needs to be improved.
image

File: src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSHostImplementation.Types.cs [hunk 1]
```diff
@@ -46,5 +46,38 @@ public struct IntPtrAndHandle
...
+
+        // keep in sync with types\internal.ts
+        public enum JSThreadInteropMode : int
+        {
+            // throw PlatformNotSupportedException if synchronous JSImport/JSExport is called on threads with JS interop, like JSWebWorker and Main thread.
+            // calling synchronous JSImport on thread pool or new threads is allowed.
+            NoSyncJSInterop = 0,
+            // allow non-re-entrant synchronous blocking calls to and from JS on JSWebWorker on threads with JS interop, like JSWebWorker and Main thread.
+            // calling synchronous JSImport on thread pool or new threads is allowed.
+            SimpleSynchronousJSInterop = 1,
```
Comment (maraf) [inline src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSHostImplementation.Types.cs:80]: I think we should come with a better name (before api review?)

File: src/mono/browser/runtime/managed-exports.ts [hunk 2]
```diff
@@ -122,8 +123,8 @@ export function release_js_owned_object_by_gc_handle(gc_handle: GCHandle) {
         set_arg_type(arg1, MarshalerType.Object);
         set_gc_handle(arg1, gc_handle);
         if (is_gcv_handle(gc_handle)) {
-            // this must stay synchronous for free_gcv_handle sake
-            invoke_sync_jsexport(managedExports.ReleaseJSOwnedObjectByGCHandle, args);
+        // this must stay synchronous for free_gcv_handle sake
```
Comment (maraf) [inline src/mono/browser/runtime/managed-exports.ts]: nit: Indentation

**Review by maraf — APPROVED** [review]

File: src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSHostImplementation.Types.cs (hunk 1 above)
Comment (pavelsavara) [reply src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSHostImplementation.Types.cs:80]: @lewing also didn't like it. Please come up with suggestions.

File: src/libraries/System.Runtime.InteropServices.JavaScript/tests/System.Runtime.InteropServices.JavaScript.UnitTests/BackgroundExec/System.Runtime.InteropServices.JavaScript.BackgroundExec.Tests.csproj [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<Project>
+  <PropertyGroup>
+    <_XUnitBackgroundExec Condition="'$(_XUnitBackgroundExec)' == ''">true</_XUnitBackgroundExec>
+    <AssemblyName>System.Runtime.InteropServices.JavaScript.Tests</AssemblyName>
```
[the remark this answers is not in the document]
Comment (pavelsavara) [reply src/libraries/System.Runtime.InteropServices.JavaScript/tests/System.Runtime.InteropServices.JavaScript.UnitTests/BackgroundExec/System.Runtime.InteropServices.JavaScript.BackgroundExec.Tests.csproj:4]: this makes the `getAssemblyExports("System.Runtime.InteropServices.JavaScript.Tests.dll")` work for both of them.

File: src/mono/browser/runtime/invoke-js.ts [hunk 4]
```diff
@@ -151,12 +152,12 @@ function bind_js_import(signature: JSFunctionSignature): Function {
 
     let wrapped_fn: WrappedJSFunction = bound_fn;
     if (WasmEnableThreads) {
-        if (is_async || is_discard_no_wait) {
-            wrapped_fn = async_bound_fn;
-        }
-        else {
-            wrapped_fn = sync_bound_fn;
-        }
+    if (is_async || is_discard_no_wait) {
```
[the remark this answers is not in the document]
Comment (pavelsavara) [reply src/mono/browser/runtime/invoke-js.ts]: fixed

File: src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSHostImplementation.Types.cs (hunk 1 above)
Comment (maraf) [reply src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSHostImplementation.Types.cs:80]: I'm voting for loger and more descriptive name, "simple" has too much meanings

Comment (maraf) [reply src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSHostImplementation.Types.cs:80]: - SingleSynchronousCallChainJSInterop
- SingleSynchronousCallJSInterop

Comment (pavelsavara) [reply src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSHostImplementation.Types.cs:80]: `SingleSynchronousCallJSInterop` is not bad, I like it

Comment (pavelsavara) [reply src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSHostImplementation.Types.cs:80]: BTW: those would not be visible on C# API, just JS config

**Pull request merged**